tests/pgtap/test_process_payload.sql

```diff
@@ -0,0 +1,264 @@
+  '2023-04-28 15:55:00+00',
+  '2023-04-28 15:55:00+00',
+  '71ba4b00-245e-4189-9ee8-0016a3ac274d',
+  'handler_foo',
+  100,
+  'PENDING',
+  null,
+  null
+);
+
+INSERT INTO swoop.thread VALUES (
+  '2023-04-28 15:58:00+00',
+  '2023-04-28 15:58:00+00',
+  '81842304-0aa9-4609-89f0-1c86819b0752',
+  'handler_foo',
+  100,
+  'SUCCESSFUL',
+  null,
+  null
+);
+
+INSERT INTO swoop.thread VALUES (
+  '2023-04-30 15:58:00+00',
+  '2023-04-30 15:58:00+00',
+  '7716319b-5064-41fc-be04-0e1330c0c290',
+  'handler_foo',
+  100,
+  'PENDING',
+  null,
+  null
+);
+
+
+SELECT results_eq(
+  $$
+    SELECT swoop.process_payload(convert_to('hash1', 'utf-8'), 2::smallint)
+  $$,
+  $$
+    VALUES (
+      '81842304-0aa9-4609-89f0-1c86819b0752'::uuid
+    )
+  $$,
+  'return action_uuid when cache, action, thread entries exist and PENDING.'
+);
+
+
+SELECT results_eq(
+  $$
+    SELECT swoop.process_payload(convert_to('hash1', 'utf-8'), 3::smallint)
+  $$,
+  $$
+    VALUES (
+      '81842304-0aa9-4609-89f0-1c86819b0752'::uuid
+    )
+  $$,
+  'return action_uuid when cache, action, thread entries exist and SUCCESSFUL.'
+);
+
+
+SELECT results_eq(
+  $$
+    SELECT swoop.process_payload(convert_to('hash1', 'utf-8'), 4::smallint)
+  $$,
+  $$
+    VALUES (
+      null::uuid
+    )
+  $$,
+  'return null when workflow_version is greater than what is in database.'
+);
+
+
+/* Set invalid_after to an arbitrary date far in the future for
+ a payload_uuid, and set the corresponding thread for the action
+associated with that payload_uuid to a FAILED state
+*/
+
+UPDATE swoop.payload_cache SET invalid_after = '2999-01-01 00:00:00+00'
+WHERE payload_uuid = 'e8d87aa6-6c42-47ed-a33c-94498fb2c20e';
+UPDATE swoop.thread SET status = 'FAILED'
+WHERE action_uuid = '81842304-0aa9-4609-89f0-1c86819b0752';
+
+SELECT results_eq(
+  $$
+    SELECT swoop.process_payload(convert_to('hash1', 'utf-8'), 3::smallint)
+  $$,
+  $$
+    VALUES (
+      null::uuid
+    )
+  $$,
+  'return null when invalid_after is not null and greater than current date.'
+);
+
+
+ROLLBACK TO SAVEPOINT action_inserted;
+
+DELETE FROM swoop.thread
+WHERE action_uuid = '7716319b-5064-41fc-be04-0e1330c0c290';
+
+
+SELECT results_eq(
+  $$
+    SELECT swoop.process_payload(convert_to('hash4', 'utf-8'), 2::smallint)
+  $$,
+  $$
+    VALUES (
+      null::uuid
+    )
+  $$,
+  'return null for payload that has cache and action entry but no thread'
+);
+
+
+ROLLBACK TO SAVEPOINT payload_inserted;
+
+SELECT results_eq(
+  $$
+    SELECT swoop.process_payload(convert_to('hash1', 'utf-8'), 2::smallint)
+  $$,
+  $$
+    VALUES (
+      null::uuid
+    )
+  $$,
+  'return null for payload that has cache but no action entry.'
+);
+
+
+ROLLBACK TO SAVEPOINT nothing_inserted;
+
+SELECT results_eq(
+  $$
+    SELECT swoop.process_payload(convert_to('hash_test', 'utf-8'), 1::smallint)
+  $$,
+  $$
+    VALUES (
+      null::uuid
+    )
+  $$,
+  'return null for payload that does not have a cache entry'
+);
```

Comment:
Why not put this test first, so you don't have to rollback to this savepoint (or have this savepoint at all)?